query/logicalplan/optimize.go

```diff
@@ -1,16 +1,81 @@
 package logicalplan
 
+import (
+	"golang.org/x/exp/slices"
+)
+
 type Optimizer interface {
 	Optimize(plan *LogicalPlan) *LogicalPlan
 }
 
 var DefaultOptimizers = []Optimizer{
+	&AverageAggregationPushDown{},
 	&PhysicalProjectionPushDown{},
 	&FilterPushDown{},
 	&DistinctPushDown{},
 	&ProjectionPushDown{},
 }
 
+type AverageAggregationPushDown struct{}
+
+func (p *AverageAggregationPushDown) Optimize(plan *LogicalPlan) *LogicalPlan {
```

Comment:
This is also something I'm not sure belongs in our optimizer code. Intuitively, optimizations are applied conditionally based on some heuristics, while this I think we always want to do so it seems like it belongs in our planning code.

Reply:
Agreed, this is where I also started back in December iirc. However, the way the structs need to be constructed would have resulted in the logical and physical plan packages having cyclic dependencies. In case we want to put it elsewhere we should rethink the packages in some cases (e.g. create another package both can rely on).

Reply:
Could you elaborate? I'm not sure I understand the cyclic dependency.